### PR TITLE
[CARBONDATA-2894] Add support for complex map type through spark carbon file format API

### PR DESCRIPTION
**Pull request by manishgupta88**
This PR supports loading querying complex map type through spark carbon file format API.

**Note: This PR is dependent on PR #2649** 

 - [ ] Any interfaces changed?
 No
 - [ ] Any backward compatibility impacted?
 No
 - [ ] Document update required?
No
 - [ ] Testing done
Added test cases       
 - [ ] For large changes, please consider breaking it into sub-tasks under an umbrella JIRA. 
NA

### CODE REVIEW

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/6426/

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/6427/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/8094/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/31/

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/6463/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/8150/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/79/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/101/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/8172/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/103/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/8174/

**Comment by manishgupta88** [comment]
retest this please

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/8179/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/108/

File: integration/spark-datasource/src/main/scala/org/apache/spark/sql/carbondata/execution/datasources/SparkCarbonFileFormat.scala [hunk 1]
```diff
@@ -214,6 +216,30 @@ class SparkCarbonFileFormat extends FileFormat
       data
     }
 
+    private def extractMapData(data: AnyRef, mapType: MapType): ArrayObject = {
+      val mapData = data.asInstanceOf[MapData]
+      val keys: ArrayData = mapData.keyArray()
+      val values: ArrayData = mapData.valueArray()
+      var keyValueHolder = scala.collection.mutable.ArrayBuffer[AnyRef]()
```
Comment (ravipesala) [inline integration/spark-datasource/src/main/scala/org/apache/spark/sql/carbondata/execution/datasources/SparkCarbonFileFormat.scala]: I think better you can directly use `extractData(row: ArrayData)` on key and value. And after that apply zip to create array of structobject

File: integration/spark-datasource/src/main/scala/org/apache/spark/sql/carbondata/execution/datasources/CarbonSparkDataSourceUtil.scala [hunk 2]
```diff
@@ -198,7 +249,20 @@ object CarbonSparkDataSourceUtil {
       val dataType = convertSparkToCarbonDataType(field.dataType)
       dataType match {
         case s: CarbonStructType =>
-          new Field(field.name, s, s.getFields)
+          val subFields = prepareSubFields(field.name, s)
```
Comment (ravipesala) [inline integration/spark-datasource/src/main/scala/org/apache/spark/sql/carbondata/execution/datasources/CarbonSparkDataSourceUtil.scala]: Why do we need to create subfields here?

Comment (manishgupta88) [reply integration/spark-datasource/src/main/scala/org/apache/spark/sql/carbondata/execution/datasources/CarbonSparkDataSourceUtil.scala]: moved the preparation to Field class

File: integration/spark-datasource/src/main/scala/org/apache/spark/sql/carbondata/execution/datasources/SparkCarbonFileFormat.scala (hunk 1 above)
Comment (manishgupta88) [reply integration/spark-datasource/src/main/scala/org/apache/spark/sql/carbondata/execution/datasources/SparkCarbonFileFormat.scala]: ok done

**Comment by manishgupta88** [comment]
@ravipesala ...handled review comments. Kindly review and merge

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/185/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/8256/

**Comment by ravipesala** [comment]
LGTM

File: store/sdk/src/main/java/org/apache/carbondata/sdk/file/Field.java [hunk 3]
```diff
@@ -213,4 +218,58 @@ public String getColumnComment() {
...
+        children = ((StructType) subFields.getDataType()).getFields();
+      }
+    }
+  }
+
+  /**
+   * prepare sub fields for complex types
+   *
+   * @param fieldName
+   * @param dType
```
Comment (jackylk) [inline store/sdk/src/main/java/org/apache/carbondata/sdk/file/Field.java]: change to dataType and add all comment

Comment (manishgupta88) [reply store/sdk/src/main/java/org/apache/carbondata/sdk/file/Field.java]: ok

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/8266/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/195/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/17/

**Comment by ravipesala** [comment]
retest this please

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/160/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.3.1, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.3/8399/

**Comment by ravipesala** [comment]
retest this please

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.3.1, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.3/8405/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/166/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/168/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.3.1, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.3/8407/

**Comment by ravipesala** [comment]
LGTM

**Pull request closed without merging**